lib/analyzer/image-extractor.ts

```diff
@@ -1,56 +1,61 @@
-import { streamToBuffer, streamToString } from "../stream-utils";
+import { streamToString } from "../stream-utils";
 
-export { extractImageKeyFiles, ExtractedImage, ExtractedKeyFiles };
+export {
+  extractFromTar,
+  ExtractedFiles,
+  ExtractFileCallback,
+  mapLookups,
+  LookupEntry,
+};
 
-interface ExtractedKeyFiles {
-  txt: { [key: string]: string };
-  md5: { [key: string]: string };
+interface ExtractedFiles {
+  [key: string]: string;
 }
 
-interface ExtractedImage {
-  manifest: string;
-  layers: Layers;
+interface ExtractedFilesByLayers {
+  [key: string]: ExtractedFiles;
 }
 
-interface Layers {
-  [key: string]: ExtractedKeyFiles;
+type ExtractFileCallback = (stream: Readable) => Promise<string>;
+
+interface LookupEntry {
+  p: string; // path pattern to look for
+  c: ExtractFileCallback; // handler to manipulate stream into a string
 }
 
+/**
+ * Create lookup entries array by mapping the specified callback with each of the specified patterns
+ * @param patterns array of path pattern strings
+ * @param callback handler to manipulate stream into a string
+ */
+const mapLookups = (patterns, callback) => {
+  return patterns.map((pattern) => {
+    return { p: pattern, c: callback };
+  });
+};
+
 /**
  * Extract key files form the specified TAR stream.
- * @param stream image layer as a Readable TAR stream
- * @param name layer name
- * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
- * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
- * @returns key files found in the specified layer TAR stream
+ * @param layerTarStream image layer as a Readable TAR stream
+ * @param lookups array of pattern, callback pairs
+ * @returns extracted file products
  */
-async function extractLayerKeyFiles(
-  stream: Readable,
-  txtPatterns: string[],
-  md5Patterns: string[] = [],
-): Promise<ExtractedKeyFiles> {
+async function extractFromLayer(
+  layerTarStream: Readable,
+  lookups: LookupEntry[],
+): Promise<ExtractedFiles> {
   return new Promise((resolve) => {
-    const result: ExtractedKeyFiles = { txt: {}, md5: {} };
+    const result: ExtractedFiles = {};
     const layerExtract: Extract = extract();
     layerExtract.on("entry", (header, stream, next) => {
-      txtPatterns.forEach((pattern) => {
-        if (minimatch(header.name, pattern, { dot: true })) {
-          streamToString(stream).then((value) => {
-            result.txt[header.name] = value;
-          });
-        }
-      });
-      md5Patterns.forEach((pattern) => {
-        if (minimatch(header.name, pattern, { dot: true })) {
-          streamToBuffer(stream).then((value) => {
-            result.md5[header.name] = md5(value);
-          });
+      lookups.forEach((lookup) => {
+        if (minimatch(`/${header.name}`, lookup.p, { dot: true })) {
+          lookup.c(stream).then((value) => (result[`/${header.name}`] = value));
```

Comment:
I think the results should be split into array according to the steam - otherwise there is any ovelap